profiler/src/ProfilerEngine/Datadog.Profiler.Native/LiveObjectsProvider.cpp

```diff
@@ -0,0 +1,223 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+
+#include <string>
+#include <vector>
+
+#include "GarbageCollection.h"
+#include "IConfiguration.h"
+#include "LiveObjectsProvider.h"
+#include "OpSysTools.h"
+#include "Sample.h"
+
+std::vector<SampleValueType> LiveObjectsProvider::SampleTypeDefinitions(
+{
+    {"inuse-objects", "count"},
+    {"inuse-space", "bytes"}
+});
+
+
+const uint32_t MAX_LIVE_OBJECTS = 1024;
+
+
+LiveObjectsProvider::LiveObjectsProvider(
+    uint32_t valueOffset,
+    ICorProfilerInfo13* pCorProfilerInfo,
+    IManagedThreadList* pManagedThreadList,
+    IFrameStore* pFrameStore,
+    IThreadsCpuManager* pThreadsCpuManager,
+    IAppDomainStore* pAppDomainStore,
+    IRuntimeIdStore* pRuntimeIdStore,
+    IConfiguration* pConfiguration)
+    :
+    _valueOffset(valueOffset),
+    _pCorProfilerInfo(pCorProfilerInfo),
+    _pFrameStore(pFrameStore),
+    _pAppDomainStore(pAppDomainStore),
+    _pRuntimeIdStore(pRuntimeIdStore),
+    _isTimestampsAsLabelEnabled(pConfiguration->IsTimestampsAsLabelEnabled())
+{
+    _pAllocationsProvider = std::make_unique<AllocationsProvider>(
+        valueOffset,  // the values (allocation count and size are stored in the live object values, not tha allocation values)
+        pCorProfilerInfo,
+        pManagedThreadList,
+        pFrameStore,
+        pThreadsCpuManager,
+        pAppDomainStore,
+        pRuntimeIdStore,
+        pConfiguration,
+        nullptr);
+}
+
+const char* LiveObjectsProvider::GetName()
+{
+    return "LiveObjectsProvider";
+}
+
+void LiveObjectsProvider::OnGarbageCollectionStart(
+    int32_t number,
+    uint32_t generation,
+    GCReason reason,
+    GCType type
+    )
+{
+    std::lock_guard<std::mutex> lock(_liveObjectsLock);
+
+    // address provided during AllocationTick event were not pointing to real objects
+    // so we have to wait for the next garbage collection to create a wrapping weak handle
+    for (auto& info : _objectsToMonitor)
+    {
+        auto handle = CreateWeakHandle(info.GetAddress());
+
+        if (handle != nullptr)
+        {
+            info.SetHandle(handle);
+        }
+        else
+        {
+            // this should never happen
+        }
+    }
```

Comment:
This can be entirely removed no?

In the `OnAllocation` method, you created the weak handle already.
Since you patch the address in `OnAllocation`, there is no need anymore of a 2 queues.

You can check if in the `_monitoredObjects` queue, there is still room. Add your object.
On garbage collector end, you process the queue.

Reply:
You're right.
I did not remove the code before ensuring that patching the MT was enough.
✅ done